src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/UseCrossPlatformIntrinsicsFixer.cs

```diff
@@ -0,0 +1,76 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.NetAnalyzers;
+
+namespace Microsoft.CodeQuality.Analyzers.Maintainability
+{
+    using static MicrosoftCodeQualityAnalyzersResources;
+    using RuleKind = UseCrossPlatformIntrinsicsAnalyzer.RuleKind;
+
+    public abstract class UseCrossPlatformIntrinsicsFixer : OrderedCodeFixProvider
+    {
+        public sealed override ImmutableArray<string> FixableDiagnosticIds { get; } = ImmutableArray.Create(UseCrossPlatformIntrinsicsAnalyzer.RuleId);
+
+        protected sealed override string CodeActionTitle => UseCrossPlatformIntrinsicsTitle;
+
+        protected sealed override string CodeActionEquivalenceKey => nameof(UseCrossPlatformIntrinsicsFixer);
+
+        protected abstract SyntaxNode CreateExclusiveOrExpression(SyntaxNode left, SyntaxNode right);
+        protected abstract SyntaxNode CreateLeftShiftExpression(SyntaxNode left, SyntaxNode right);
+        protected abstract SyntaxNode CreateRightShiftExpression(SyntaxNode left, SyntaxNode right);
+        protected abstract SyntaxNode? CreateUnsignedRightShiftExpression(SyntaxNode left, SyntaxNode right);
+
+        protected sealed override Task FixAllCoreAsync(SyntaxEditor editor, SyntaxGenerator generator, Diagnostic diagnostic, CancellationToken cancellationToken)
+        {
+            SyntaxNode node = editor.OriginalRoot.FindNode(diagnostic.Location.SourceSpan, getInnermostNodeForTie: true);
+
+            // We shouldn't get here for a diagnostic that doesn't have the expected properties.
+
+            if (!diagnostic.Properties.TryGetValue(nameof(RuleKind), out string? ruleKindName))
+            {
+                Debug.Fail($"Found diagnostic without an associated {nameof(RuleKind)} property.");
```

Comment:
fwiw, we just don't write this way,a nd practically dont' find it valuable as nothing ever runs this stuff in debug *except* tests.  and in that case, tests are already checking that things are working or not.

Reply:
That isn't the only place, we also get various coverage in source-build and it is very beneficial when you hit an analyzer failure that you're trying to diagnose, as you can just bin-place the debug version of the analyzer into your NuGet cache.

We also regularly find cases where some code is taking advantage of a new language feature and hits a throwing path in an analyzer, such as a bad assumption being made that `IParameterSymbol` are only found as children of methods (something found in the data flow analyzer for the IL linker recently).

These asserts help clarify there is an expected invariant and help quickly identify where to start looking when a failure is found.